### PR TITLE
Add Windows builds

### PR DESCRIPTION
**Pull request by whytro**
> Related to #11 

Changes made:

- Added Windows-related CI.
- Removed prebuilt tests (users will now have to run the make file or bat file in the test/data dir).
- Changed `__main__` to search in more directories, and also look for `osrm-datastore` or `osrm-datastore.exe` (since the previous default sitepackages dir was different on Windows unless the `--user` flag was specified).
- Shifted artifact downloading to its own action, and moved the rest of the `before-all` to the pyproject.toml file to cut down on code duplication.
- Added basic .gitignore file.
- Adjusted a test to no longer require a specific order of returned valid options, as it was dependent on map order.
- Executable calls (ie. `osrm-datastore`, and the test generation make/bat) have been moved to the `test-command` section, as it requires the package to be installed due to reliance on `python -m osrm <executable>`.
- Fixed execute permissions being missing on Linux/Mac builds.

---
Since the osrm-backend artifact workflow for Windows builds store the dependencies, building the actual wheel for Windows doesn't take too long - about 7~9 minutes. The osrm-backend artifact workflow for Windows takes like 40 minutes though.
Should we enable the Windows wheel builds on PR CI runs, or leave it out since it's still 2x/3x the time taken compared to Mac or Linux?

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -85,7 +85,9 @@ if(LibOSRM_FOUND)
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${LibOSRM_CXXFLAGS}")
 
   file(GLOB OSRM_EXECS ${LibOSRM_EXEC_DIR}/osrm-*)
-  file(COPY ${OSRM_EXECS} DESTINATION ${CMAKE_CURRENT_SOURCE_DIR}/src/bin)
+  install(FILES ${OSRM_EXECS} TYPE BIN 
+          PERMISSIONS OWNER_WRITE OWNER_READ GROUP_READ WORLD_READ 
+          OWNER_EXECUTE GROUP_EXECUTE WORLD_EXECUTE)
```
Comment (nilsnolde) [inline CMakeLists.txt:90]: so this worked now with permissions?

File: src/osrm/__main__.py [hunk 2]
```diff
@@ -6,7 +7,21 @@
...
-exec = site.getusersitepackages()+"/bin/"
+searchpaths = site.getsitepackages()
+if(site.ENABLE_USER_SITE):
+    searchpaths.append(site.getusersitepackages())
+
+exec = ""
+
+for path in searchpaths:
+    currpath = path+"/bin/"
+    if os.path.isfile(currpath+"osrm-datastore") or os.path.isfile(currpath+"osrm-datastore.exe"):
```
Comment (nilsnolde) [inline src/osrm/__main__.py:18]:
```suggestion
    currpath = path + "/bin/"
    if os.path.isfile(currpath + "osrm-datastore") or os.path.isfile(currpath + "osrm-datastore.exe"):
```

**Comment by nilsnolde** [comment]
This is awesome @whytro , very good job! 

> Should we enable the Windows wheel builds on PR CI runs, or leave it out since it's still 2x/3x the time taken compared to Mac or Linux?

I'd suggest the following for CI:
- in the PR one enable only linux, test all python versions, this is quick
- in master_push, test only 3.8, but all platforms, takes longer but is much less used

WDYT?

After we merged this, we can try another publishing to test PyPI. Once that's successful, we can switch to the real thing and it's a v0.0.1 for real:) That's a really good end of the project.

File: CMakeLists.txt (hunk 1 above)
Comment (whytro) [reply CMakeLists.txt:90]: Yep, this change allowed the executables to have execute permissions!

File: src/osrm/__main__.py [hunk 3]
```diff
@@ -6,7 +7,21 @@
...
+        exec = currpath
+        break
```
[the remark this answers is not in the document]
Comment (whytro) [reply src/osrm/__main__.py:20]: `getsitepackages()` returns the `pip install . --user` directory on Windows. On Windows, it seems that `pip install .` installs it to a path in `getsitepackages()` (both paths are different). As a result, both paths are searched.

File: pyproject.toml [hunk 4]
```diff
@@ -21,19 +21,56 @@ wheel.py-api = "cp312"
...
+
+  export HOMEBREW_NO_AUTO_UPDATE=1 HOMEBREW_NO_INSTALL_CLEANUP=1
+  brew install lua tbb boost@1.76
+  brew link boost@1.76
+"""
+
+[tool.cibuildwheel.windows]
+before-all = [
+  "tar -zxf osrm-backend.tgz",
+  "((robocopy /E /IS /NFL /NDL /NJH /NJS /NP /MOVE .conan C:\\.conan) ^& IF %ERRORLEVEL% LSS 8 SET ERRORLEVEL = 0)",
```
[the remark this answers is not in the document]
Comment (nilsnolde) [reply pyproject.toml:65]: OMG these sorry excuses of "shells" in windows..

File: src/osrm/__main__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nilsnolde) [reply src/osrm/__main__.py]: ouch, thanks for the explanation

**Comment by nilsnolde** [comment]
Is this ok for you @SiarheiFedartsou ?

**Review by SiarheiFedartsou — APPROVED** [review]

**Pull request merged**